File: app/src/main/java/com/blackmirror/dongda/presenter/HomePresenter.java
package com.blackmirror.dongda.presenter;

import com.blackmirror.dongda.domain.Interactor.HomeUseCase;
import com.blackmirror.dongda.domain.Interactor.LikePopUseCase;
import com.blackmirror.dongda.domain.Interactor.LikePushUseCase;
import com.blackmirror.dongda.domain.model.BaseDataBean;
import com.blackmirror.dongda.domain.model.HomepageDomainBean;
import com.blackmirror.dongda.domain.model.LikePopDomainBean;
import com.blackmirror.dongda.domain.model.LikePushDomainBean;
import com.blackmirror.dongda.ui.activity.HomeActivity.HomeContract;
import com.blackmirror.dongda.utils.AppConstant;
import com.blackmirror.dongda.utils.LogUtils;

import javax.inject.Inject;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Create By Ruge at 2018-05-11
 */
public class HomePresenter implements HomeContract.HomeBasePresenter {

    @Inject
    HomeUseCase homeUseCase;

    @Inject
    LikePushUseCase pushUseCase;

    @Inject
    LikePopUseCase popUseCase;

    private HomeContract.HomeView view;

    @Inject
    public HomePresenter(HomeContract.HomeView view) {
        this.view = view;
    }

    @Override
    public void getHomePageData() {
        homeUseCase.getHomePageData()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<HomepageDomainBean>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(HomepageDomainBean bean) {
                        LogUtils.d("getHomePageData onNext");
                        if (view == null){
                            return;
                        }
                        if (bean.isSuccess){
                            view.onGetHomePageData(bean);
                        }else {
                            view.onGetHomeDataError(bean);
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        LogUtils.e(HomePresenter.class,e);
                        if (view == null){
                            return;
                        }
                        view.onGetHomeDataError(getErrorData(e));

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public void likePush(String service_id) {
        pushUseCase.likePush(service_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<LikePushDomainBean>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(LikePushDomainBean bean) {
                        if (view == null){
                            return;
                        }
                        if (bean.isSuccess){
                            view.onLikePushSuccess(bean);
                        }else {
                            view.onGetHomeDataError(bean);
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        LogUtils.e(HomePresenter.class,e);
                        if (view == null){
                            return;
                        }
                        view.onGetHomeDataError(getErrorData(e));

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public void likePop(String service_id) {
        popUseCase.likePop(service_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<LikePopDomainBean>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(LikePopDomainBean bean) {
                        if (view == null){
                            return;
                        }
                        if (bean.isSuccess){
                            view.onLikePopSuccess(bean);
                        }else {
                            view.onGetHomeDataError(bean);
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        LogUtils.e(HomePresenter.class,e);
                        if (view == null){
                            return;
                        }
                        view.onGetHomeDataError(getErrorData(e));

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    private BaseDataBean getErrorData(Throwable e) {
        BaseDataBean bean = new BaseDataBean();
        bean.code = AppConstant.NET_UNKNOWN_ERROR;
        bean.message = e.getMessage();
        return bean;
    }

    @Override
    public void setView(HomeContract.HomeView view) {

    }

    @Override
    public void destroy() {

    }
}

File: app/src/main/java/com/blackmirror/dongda/presenter/HomePresenter.kt
package com.blackmirror.dongda.presenter


import com.blackmirror.dongda.kdomain.interactor.likePopImpl
import com.blackmirror.dongda.kdomain.interactor.likePushImpl
import com.blackmirror.dongda.kdomain.interactor.searchServiceImpl
import com.blackmirror.dongda.kdomain.model.BaseDataBean
import com.blackmirror.dongda.ui.activity.homeActivity.HomeContract
import com.blackmirror.dongda.utils.AppConstant
import com.blackmirror.dongda.utils.logE
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

/**
 * Create By Ruge at 2018-05-11
 */
class HomePresenter @Inject constructor(val view: HomeContract.HomeView?) : HomeContract.HomeBasePresenter {

    override fun getHomePageData() {
        searchServiceImpl()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it.isSuccess) {
                        view?.onGetHomePageData(it)
                    } else {
                        view?.onGetHomeDataError(it)
                    }
                },{
                    view?.onGetHomeDataError(getErrorData(it))
                })
    }

    override fun likePush(service_id: String) {
        likePushImpl(service_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({if (it.isSuccess) {
                    view?.onLikePushSuccess(it)
                } else {
                    view?.onGetHomeDataError(it)
                }},{
                    logE(message = HomePresenter::class.java.simpleName, exception = it)
                    view?.onGetHomeDataError(getErrorData(it))
                })
    }

    override fun likePop(service_id: String) {
        likePopImpl(service_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it.isSuccess) {
                        view?.onLikePopSuccess(it)
                    } else {
                        view?.onGetHomeDataError(it)
                    }
                },{e->
                    logE(message = HomePresenter::class.java.simpleName, exception = e)
                    view?.onGetHomeDataError(getErrorData(e))
                })
    }

    private fun getErrorData(e: Throwable): BaseDataBean {
        val bean = BaseDataBean()
        bean.code = AppConstant.NET_UNKNOWN_ERROR
        bean.message = e.message
        return bean
    }

    override fun setView(view: HomeContract.HomeView) {

    }

    override fun destroy() {

    }
}
